src/Agent.Plugins/PipelineCache/TarUtils.cs

```diff
@@ -0,0 +1,202 @@
+                    }
+                    ExceptionDispatchInfo.Capture(e).Throw();
+                }
+            }
+            return archiveFile;
+        }
+
+        public async Task<bool> DownloadTar(AgentTaskPluginExecutionContext context, DedupManifestArtifactClient dedupManifestClient, DedupIdentifier dedupId, string targetDirectory, CancellationToken cancellationToken)
+        {
+            var processTcs = new TaskCompletionSource<int>();
+
+            using (var cancelSource = new CancellationTokenSource())
+            using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cancelSource.Token))
+            using (var process = new Process())
+            {
+                process.StartInfo.FileName = isWindows ? "7z" : "tar";
+                process.StartInfo.Arguments = isWindows ? $"x -si -aoa -o{targetDirectory} -ttar" : $"-xf - -C {targetDirectory}";
+                process.StartInfo.UseShellExecute = false;
+                process.StartInfo.RedirectStandardInput = true;
+                process.StartInfo.RedirectStandardOutput = true;
+                process.StartInfo.RedirectStandardError = true;
+                process.EnableRaisingEvents = true;
+                process.Exited += (sender, args) =>
+                {
+                    cancelSource.Cancel();
+                    processTcs.SetResult(process.ExitCode);
+                };
+
+                context.Info($"Starting '{process.StartInfo.FileName}' with arguments '{process.StartInfo.Arguments}'...");
+                process.Start();
+
+                var output = new List<string>();
+
+                Task readLines(string prefix, StreamReader reader) => Task.Run(async () =>
+                  {
+                      string line;
+                      while (null != (line = await reader.ReadLineAsync()))
+                      {
+                          lock (output)
+                          {
+                              output.Add($"{prefix}{line}");
+                          }
+                      }
+                  });
+
+                Task readStdOut = readLines("stdout: ", process.StandardOutput);
+                Task readStdError = readLines("stderr: ", process.StandardError);
+                Task downloadTask = Task.Run(async () =>
+                {
+                    try
+                    {
+                        await dedupManifestClient.DownloadToStreamAsync(dedupId, process.StandardInput.BaseStream, proxyUri: null, linkedSource.Token);
+                        process.StandardInput.BaseStream.Close();
+                    }
+                    catch (Exception e)
+                    {
+                        process.Kill();
+                        ExceptionDispatchInfo.Capture(e).Throw();
+                    }
+                });
+
+                // Our goal is to always have the process ended or killed by the time we exit the function.
+                try
```

Comment:
Feels like a lot of duplicated logic between this and the upload method ... both methods are essetially ding the same thing (running a process), checking for certain exit codes, waiting on certain things, handling exceptions., etc ... feels like you could move this into a common method and avoid a lot of the duplication.